### DISCUSSION TITLE
Is there a way to pause the script every time a variable changes?

### DISCUSSION THREAD

**Discussion by Wilson703** [zero-plusplus/vscode-autohotkey-debug#334] [Q&A]
Some debuggers on the VS code have the ability to pause on a variable change. Is that something that can be achieved with any of the functionalities of vscode-autohotkey-debug?

**Comment by zero-plusplus** [comment]
In conclusion, it is very difficult to implement a [data breakpoint](https://code.visualstudio.com/Docs/editor/debugging#_data-breakpoints) that breaks when data is changed only the debugger adapter side.

I previously presented an idea for implementing it on the debugger adapter side [here](https://github.com/zero-plusplus/vscode-autohotkey-debug/issues/93#issuecomment-797852841), but I have not been able to implement it to date because it would be very complex to implement and would likely be incomplete in terms of functionality.

**Comment by yaronelh** [comment]
Thanks for the explanation, it would definitely be extremely useful if it was possible

**Reply by zero-plusplus** [reply]
Just out of curiosity, can you tell me in what cases people are using data breakpoints?

And I would like to consider whether it can be solved by the implementation of the above proposal.

I don't use this feature very often, I thought it would be useful to have this feature when I was doing some kind of testing before but forgot about it.

Instead of using data breakpoints, I have currently been able to solve most cases by referencing variables through static analysis or using conditional breakpoints.

**Comment by Wilson703** [comment]
I don't know what other people will use it for, but for me I thought of using it without any breakpoints, just to follow a variable through its life cycle and see if at a certain point it either receives the wrong value, or unexpected variable and where it happens in the code 

it would be way more time-consuming to start placing breakpoints throughout the code and trying to figure that out. You just declare a variable to follow and see how it progresses, stopping the script when it changes so I can see what's happening.

The practical advantage is not needing to declare a breakpoint, just having that happen automatically.

**Reply by SAbboushi** [reply]
Something I've wanted as well: found that very useful in other debuggers (not related to AHK).

@zero-plusplus: when we dialoged about this previously [in link above](https://github.com/zero-plusplus/vscode-autohotkey-debug/issues/93#issuecomment-797852841), you said:
"The second reason is that the AutoHotkey debugger does not notify data changes."
I'm hoping @Lexikos will add his comments on this matter.

**Reply by zero-plusplus** [reply]
Thanks, I understand the benefits now.

I remembered thinking that this feature would be useful in unit tests, as I often only focus on specific variables, `expected` and `actual`.
However, this is a usage related to the break with watch expression suggested in the link above, which seems to be slightly different from the benefits of data breakpoints. (It can be considered a variant of data breakpoints in that it targets a specific variable name regardless of its specific scope, such as a function.)

Since the reliability of determining whether a value has changed is extremely important for this feature, it would be preferable to have this resolved on the AutoHotkey Debugger side rather than a complex implementation on the debugger adapter side.

If you want them to implement it on the AutoHotkey debugger side, they need to understand the convenience they will get by implementing it for Lexikos.
It may be a good idea to provide many more concrete examples of use.

---

Unfortunately, I am no longer maintaining this extension, so I will not be able to immediately support it even if it is implemented in the AutoHotkey debugger.

I plan to restructure the debugger adapter in a [new extension](https://github.com/zero-plusplus/autohotkey-devtools), but due to the scale of the project, I think it will take several years.

**Comment by Wilson703** [comment]
Thanks for the reply, I'm still using v1 so I don't think it would be relevant even if @Lexikos decides to implement this. and if you're no longer developing the debugger, that would make it even less likely to ever be implemented

**Reply by zero-plusplus** [reply]
I forgot that no new features will be added to v1.

I would like to fully support v1 in the new extension, so it would be interesting to implement this on the debugger adapter side.
I can implement is pseudo data breakpoints, so they will be unreliable and it will take time to implement, so I can't promise anything, but I left a note [here](https://github.com/zero-plusplus/autohotkey-devtools/issues/2#pseudo-data-breakpoints) so I don't forget.
